### PR TITLE
Possibility to align minimized MDI child to default location (TOP LEFT)

### PR DESCRIPTION
**Pull request by kirsan31**
Fix https://github.com/dotnet/winforms/issues/4720.

## Proposed changes

- Add public bool property `MdiChildrenMinimizedAnchorBottom` to `Form` class.

## Customer Impact

WinForms override standard MDI behavior without option to use original one. It's lead to all child windows layout (with minimized windows) are mess up after main window resize. This can happened after simple resize, or after resolution changed (for example if your connect by rdp from other device), etc... Customer can use `MdiChildrenMinimizedAnchorBottom` to revert for Windows default behavior.

## Regression? 

No

## Risk

Minimal.

## Screenshots
Designer

https://user-images.githubusercontent.com/17767561/125053862-3e5ac700-e0ae-11eb-8ff1-3b5f72500f9d.mp4

P.s. Video above with old name of the property.
## Test methodology

Manual.
Maui automated test.

## Accessibility testing  

Unfortunately, I do not know anything on this topic. Therefore, if this is necessary, I will need additional time to figure it out ...

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/5221)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=5221) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=5221) before we can accept your contribution.<br/><br/>:x: kirsan31 [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=5221)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=5221) it.</sub>

**Comment by kirsan31** [comment]
> @kirsan31, Good news - API is approved with suggestions. Details are in the associated issue. Can you please update this PR and mark it ready for review?

Excellent. I'll do it as soon as I can.

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 1]
```diff
@@ -3126,6 +3126,9 @@ Do you want to replace it?</value>
   <data name="FormMDIChildrenDescr" xml:space="preserve">
     <value>Retrieves the MDI children of this form.</value>
   </data>
+  <data name="FormMdiChildrenMinimizedAnchorBottomDescr" xml:space="preserve">
+    <value>true - anchoring minimized MDI children to the BottomLeft of the parent form (WinForms default). false - anchoring minimized MDI children to the TopLeft of the parent form (windows default).</value>
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/Resources/SR.resx:3130]: I would change this description to something like, "Anchoring minimized MDI children to the BottomLeft of the parent Form (WinForms default). Disabling this flag will result in Windows default (i.e. TopLeft)"

**Review by dreddy-work — COMMENTED** [review]
@Tanya-Solyanik

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 2]
```diff
@@ -1516,6 +1516,14 @@ public Form[] MdiChildren
             }
         }
 
+        /// <summary>
+        /// true - anchoring minimized MDI children to the BottomLeft of the parent form (WinForms default). false - anchoring minimized MDI children to the TopLeft of the parent form (windows default).
+        /// </summary>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:1521]: Please have a look how all other properties are described. The summary wording must start with something like "Gets or sets...", and it mustn't contain value's information.


```suggestion
        /// <summary>
        /// Gets or sets the anchoring for minimized MDI children.
        /// </summary>
        /// <value><see langword="true" /> to anchor minimized MDI children to the bottom left of the parent form; <see langword="false" /> to anchor to the top left of the parent form (Windows default).</value>
```

File: src/System.Windows.Forms/src/Resources/SR.resx (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/Resources/SR.resx:3130]:
```suggestion
    <value>Gets or sets the anchoring for minimized MDI children.</value>
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 3]
```diff
@@ -1516,6 +1516,14 @@ public Form[] MdiChildren
...
+        [SRCategory(nameof(SR.CatWindowStyle))]
+        [DefaultValue(true)]
+        [SRDescription(nameof(SR.FormMdiChildrenMinimizedAnchorBottomDescr))]
+        public bool MdiChildrenMinimizedAnchorBottom { get; set; } = true;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:1525]: Default `true` is this consistent with the behaviour in .NET Framework?

**Review by RussKie — CHANGES_REQUESTED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs (hunk 3 above)
Comment (kirsan31) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:1525]: Yes - BottomLeft.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs (hunk 2 above)
Comment (kirsan31) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:1521]: Got it, thanks.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 4]
```diff
@@ -1516,6 +1516,15 @@ public Form[] MdiChildren
             }
         }
 
+        /// <summary>
+        /// Gets or sets the anchoring for minimized MDI children.
+        /// </summary>
+        /// <value><see langword="true" /> to anchor minimized MDI children to the bottom left of the parent form (WinForms default); <see langword="false" /> to anchor to the top left of the parent form (Windows default).</value>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:1522]:
```suggestion
        /// <value><see langword="true" /> to anchor minimized MDI children to the bottom left of the parent form; <see langword="false" /> to anchor to the top left of the parent form.</value>
```

The platform defaults need to be placed under `remarks`, e.g. https://docs.microsoft.com/dotnet/api/system.windows.forms.mdiclient.backgroundimage
Something like:
```xml
<remarks>
By default Windows Forms anchors MDI children to the bottom left of the parent form, whilst the Windows default is top left.
</remarks>
```

**Review by RussKie — CHANGES_REQUESTED** [review]
How can we test the change?

1. Please update our manual [MDI sample](https://github.com/dotnet/winforms/blob/e0c821b17863458611e4e056c16eb1ad65f9ce9a/src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/MainForm.cs#L116-L119) in a way that we can toggle the property on and off and see the result.
2. We'll also need an automated test, have a look at our maui tests and add a new test project for MDI. Maybe [MauiPrintPreviewDialogTests](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiPrintPreviewDialogTests/MauiPrintPreviewDialogTests.cs) be a good start?

**Comment by RussKie** [comment]
Rebase is generally better than tracking merges
image

**Comment by kirsan31** [comment]
> How can we test the change?
> 
> 1. Please update our manual [MDI sample](https://github.com/dotnet/winforms/blob/e0c821b17863458611e4e056c16eb1ad65f9ce9a/src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/MainForm.cs#L116-L119) in a way that we can toggle the property on and off and see the result.
> 2. We'll also need an automated test, have a look at our maui tests and add a new test project for MDI. Maybe [MauiPrintPreviewDialogTests](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiPrintPreviewDialogTests/MauiPrintPreviewDialogTests.cs) be a good start?

I will look into it...

**Comment by kirsan31** [comment]
> 1. Please update our manual [MDI sample](https://github.com/dotnet/winforms/blob/e0c821b17863458611e4e056c16eb1ad65f9ce9a/src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/MainForm.cs#L116-L119) in a way that we can toggle the property on and off and see the result.

@RussKie designer not working with this error:
image
Any advice?

**Comment by RussKie** [comment]
Hmm... We don't use assembly, at least directly anywhere... 
"Works on my machine" 😃 Though I'm on 17preview2
image


Maybe try deleting artifacts folder, then .\build.cmd, followed by .\start-vs.cmd?

**Comment by kirsan31** [comment]
> Maybe try deleting artifacts folder, then .\build.cmd, followed by .\start-vs.cmd?

Not helped sadly :( So I have 2 options here:
 - Uninstall 16.11 and install 17.
 - Manually edit `MdiChild.Designer.cs`.
 ...

**Comment by RussKie** [comment]
I'm sure it worked in v16 as well.
There's a 3rd option as well - make a quick project copy these files on it, make changes, copy back.

I don't think the extent of changes that require the designer is significant.

**Comment by kirsan31** [comment]
> I'm sure it worked in v16 as well.
> There's a 3rd option as well - make a quick project copy these files on it, make changes, copy back.
> 
> I don't think the extent of changes that require the designer is significant.

I've already install vs17p2 and it's worked :)
image
Strange...

**Comment by dreddy-work** [comment]
@kirsan31 , for VS 2022 crash, was it simple scenario of opening designer from your test application? or did you perform any other actions?

**Comment by kirsan31** [comment]
> @kirsan31 , for VS 2022 crash, was it simple scenario of opening designer from your test application? or did you perform any other actions?

@dreddy-work 
I was able normally open `src\System.Windows.Forms\tests\IntegrationTests\WinformsControlsTest\MdiChild.cs` in VS2022 (in VS2019 it was failing with error above).
But VS2022 was not displaying control's properties. And when I was trying to see some of them (selecting different controls on the form and click properties menu), VS crashed a lot. It was crashed also when I was trying to select control in this dropdown:
image

**Comment by RussKie** [comment]
It may be that the latest .NET requires the newer version...
________________________________
From: kirsan31 ***@***.***>
Sent: Thursday, 15 July 2021 11:51 PM
To: dotnet/winforms ***@***.***>
Cc: Igor Velikorossov ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [dotnet/winforms] Possibility to align minimized MDI child to default location (TOP LEFT) (#5221)


I'm sure it worked in v16 as well.
There's a 3rd option as well - make a quick project copy these files on it, make changes, copy back.

I don't think the extent of changes that require the designer is significant.

I've already install vs17p2 and it's worked :)
[image]<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F17767561%2F125799291-70bcd219-bd37-4157-8363-4bd32391f4f1.png&data=04%7C01%7Cigor.velikorossov%40microsoft.com%7Cc89ded3c3d654a7e0d6608d94797a244%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637619538879074684%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=RzHXZJM8kPuNPzsFp1obvAcRtLc17H2fJK3hbnI%2FGuA%3D&reserved=0>
Strange...

**Comment by kirsan31** [comment]
> 1. Please update our manual MDI sample in a way that we can toggle the property on and off and see the result.

Done:

https://user-images.githubusercontent.com/17767561/125908938-4e7904ca-9667-4490-b014-c7ed39eb036b.mp4

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            MdiChildrenMinimizedAnchorBottom = true;
+
+            var frm = new Form
+            {
+                MdiParent = this,
+                WindowState = FormWindowState.Minimized
+            };
+            frm.Show();
+
+            int Posdiff = ClientSize.Height - frm.Top;
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:40]: 1. Give variables meaningful name, e.g. `childFormMinimizedPositionY`
2. Local variable names must be in camelCase.

**Review by RussKie — CHANGES_REQUESTED** [review]
Almost there! 👍

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            Thread.CurrentThread.SetCulture("en-US");
+            Application.Run(new MauiMDITests(args));
+        }
+
+        [Scenario(true)]
+        public ScenarioResult Resize_When_MdiChildrenMinimizedAnchorBottom_True(TParams p)
+        {
+            MdiChildrenMinimizedAnchorBottom = true;
+
+            var frm = new Form
```
[the remark this answers is not in the document]
Comment (kirsan31) [reply src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:33]: I don't know your Maui tests infrastructure. Is it safe to call `Dispose` on child form before it was closed? Or is it just necessary?

**Review by RussKie — APPROVED** [review]

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                WindowState = FormWindowState.Minimized
+            };
+            childForm.Show();
+
+            int childFormMinimizedYPositionFromBottom = ClientSize.Height - childForm.Top;
+            Height += 100;
+
+            if (childFormMinimizedYPositionFromBottom != ClientSize.Height - childForm.Top)
+            {
+                return new ScenarioResult(false, $"MDI child changed bottom relative position with {nameof(MdiChildrenMinimizedAnchorBottom)} == true!", p.log);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:45]:
```suggestion
                return new ScenarioResult(false, $"MDI child changed bottom relative position with {nameof(MdiChildrenMinimizedAnchorBottom)} == {MdiChildrenMinimizedAnchorBottom}!", p.log);
```

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                WindowState = FormWindowState.Minimized
+            };
+            childForm.Show();
+
+            int childFormMinimizedTop = childForm.Top;
+            Height += 100;
+
+            if (childFormMinimizedTop != childForm.Top)
+            {
+                return new ScenarioResult(false, $"MDI child changed top relative position with {nameof(MdiChildrenMinimizedAnchorBottom)} == false!", p.log);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:68]:
```suggestion
                return new ScenarioResult(false, $"MDI child changed top relative position with {nameof(MdiChildrenMinimizedAnchorBottom)} == {MdiChildrenMinimizedAnchorBottom}!", p.log);
```

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        public static void Main(string[] args)
+        {
+            Thread.CurrentThread.SetCulture("en-US");
+            Application.Run(new MauiMDITests(args));
+        }
+
+        [Scenario(true)]
+        public ScenarioResult Resize_When_MdiChildrenMinimizedAnchorBottom_True(TParams p)
```
Comment (dreddy-work) [inline src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:29]:
```suggestion
        public ScenarioResult Resize_When_MdiChildrenMinimizedAnchorBottom_Deafult(TParams p)
```

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            MdiChildrenMinimizedAnchorBottom = true;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:31]: This is default value.

Comment (kirsan31) [reply src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:31]: Yes, but as I understand currently Maui test scenarios run on the same instance. And it seems that order of scenarios the same as in code. So, if I simple remove this line `MdiChildrenMinimizedAnchorBottom = true;` the test will fail. Overall I think that rely on position in code it's not right thing to do. Do you want me to use reflection to get `DefaultValueAttribute` here?

Comment (dreddy-work) [reply src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:31]: Nope. In this case, I would add this comment there for reference.

Comment (dreddy-work) [reply src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:31]: Did you forget to add comment?

Comment (kirsan31) [reply src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:31]: Ahhh, I misunderstand your again lol :)))  Iwas thinking that this is your comment to explain method rename :man_facepalming:

Comment (kirsan31) [reply src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:32]: I hope it's done on second try 😊

File: src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        public static void Main(string[] args)
+        {
+            Thread.CurrentThread.SetCulture("en-US");
+            Application.Run(new MauiMDITests(args));
+        }
+
+        [Scenario(true)]
+        public ScenarioResult Resize_When_MdiChildrenMinimizedAnchorBottom_Default(TParams p)
+        {
+            // This is default value.
```
Comment (dreddy-work) [inline src/System.Windows.Forms/tests/IntegrationTests/MauiTests/MauiMDITests/MauiMDITests.cs:31]:
```suggestion
            // Currently Maui test scenarios run on the same instance and this flag might have been set to a non-default values in the previous tests. So, explicitly setting it to `default' value `true` to make tests reliable.
```

**Review by dreddy-work — APPROVED** [review]

**Comment by RussKie** [comment]
> But VS2022 was not displaying control's properties. And when I was trying to see some of them (selecting different controls on the form and click properties menu), VS crashed a lot. It was crashed also when I was trying to select control in this dropdown:
> image

@kirsan31 have you reported those via the VS Feedback? We're actively working through the user feedback to ensure things like that don't happen.

@merriemcgaw @JeremyKuhne FYI

**Comment by kirsan31** [comment]
@RussKie 

> @kirsan31 have you reported those via the VS Feedback? We're actively working through the user feedback to ensure things like that don't happen.

Sadly no, at that moment I was frantically trying to find a way to open your form in the designer to write a test :)
P.s. In the end I wrote everything manually anyway :)
P.p.s. Your suggestion

> There's a 3rd option as well - make a quick project copy these files on it, make changes, copy back. 

naturally worked, but after small trivial change and save it in designer, whole `MdiChild.Designer.cs` was totally messed up with lots of unnecessary changes :(

**Pull request merged**